flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/utils/SavepointTestBase.java

```diff
@@ -56,16 +64,33 @@
         try {
             JobID jobID = client.submitJob(jobGraph).get();
 
-            return CompletableFuture.runAsync(waitingFunction::await)
-                    .thenCompose(ignore -> triggerSavepoint(client, jobID))
-                    .get(5, TimeUnit.MINUTES);
+            waitForAllRunningOrSomeTerminal(jobID);
+
+            return triggerSavepoint(client, jobID).get(5, TimeUnit.MINUTES);
         } catch (Exception e) {
             throw new RuntimeException("Failed to take savepoint", e);
         } finally {
             client.cancel(jobId);
         }
     }
 
+    private void waitForAllRunningOrSomeTerminal(JobID jobID) throws Exception {
+        while (true) {
```

Comment:
Should this have some sort of timeout? Or are we happy to let the CI kill the process after some time?